File: backend/app/schemas/weather.py
"""Weather-specific Pydantic schemas."""

import uuid

from pydantic import BaseModel


class CurrentWeatherResponse(BaseModel):
    """Normalized current conditions from Open-Meteo."""

    temperature: float | None = None
    apparent_temperature: float | None = None
    humidity: float | None = None
    precipitation_mm: float | None = None
    wind_speed_kmh: float | None = None
    wind_direction: str | None = None  # compass, e.g. "NW"
    conditions: str | None = None  # WMO-mapped, e.g. "Rain"
    latitude: float
    longitude: float


class ForecastDay(BaseModel):
    """A single day of forecast/historical weather."""

    date: str  # YYYY-MM-DD
    weather_code: int | None = None
    conditions: str | None = None
    temp_max: float | None = None
    temp_min: float | None = None
    precipitation_probability: int | None = None
    precipitation_sum: float | None = None
    wind_speed_max: float | None = None


class ForecastResponse(BaseModel):
    """Daily forecast or historical weather for a location."""

    latitude: float
    longitude: float
    days: list[ForecastDay]


class ActivityWeatherResponse(BaseModel):
    """Stored weather snapshot for a single activity (no external calls)."""

    activity_id: uuid.UUID
    temperature: float | None = None
    conditions: str | None = None
    wind_speed_kmh: float | None = None
    wind_direction: str | None = None
    precipitation_mm: float | None = None


class TagActivityResponse(BaseModel):
    """Result of tagging an activity with historical weather."""

    tagged: bool
    weather: ActivityWeatherResponse | None = None

File: frontend/src/lib/api/types/weather.ts
// ─── Weather ─────────────────────────────────────────────────────────────────

export interface CurrentWeather {
  temperature: number;
  apparent_temperature: number;
  humidity: number;
  precipitation_mm: number;
  wind_speed_kmh: number;
  wind_direction: string;
  conditions: string;
  latitude: number;
  longitude: number;
}

export interface ForecastDay {
  date: string; // YYYY-MM-DD
  weather_code: number;
  conditions: string;
  temp_max: number;
  temp_min: number;
  precipitation_probability: number | null;
  precipitation_sum: number;
  wind_speed_max: number;
  wind_direction_deg: number | null;
  humidity_mean: number | null;
  pressure_mean: number | null;
}

export interface ForecastResponse {
  latitude: number;
  longitude: number;
  days: ForecastDay[];
}

export interface ActivityWeather {
  temperature: number | null;
  conditions: string | null;
  wind_speed_kmh: number | null;
  wind_direction: string | null;
  wind_direction_deg: number | null;
  humidity_pct: number | null;
  pressure_hpa: number | null;
  precipitation_mm: number | null;
}
